telemetry/recover_middleware.go

```diff
@@ -0,0 +1,26 @@
+package telemetry
+
+import (
+	"fmt"
+	"net/http"
+	"runtime"
+)
+
+// RecoverMiddleware catches panics and logs them to a span.
+func RecoverMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			span := SpanFromContext(r.Context())
+			if !span.IsRecording() {
```

Comment:
should we always recover and log even if there is no span?

Reply:
If there's no span then there's no logger either (since I tightly coupled the two concepts). We could pull the logger manually and log but that doesn't buy us anything because if we don't have a span then we aren't really using telemetry anyhow.

Also, the reason to continue the panic is because Go already recovers and prints the panic for us: https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/net/http/server.go;l=1822.

So either we catch it here and log it with otel or we just let Go do its thing.